lib/segment/src/index/field_index/full_text_index/inverted_index.rs

```diff
@@ -20,47 +20,77 @@ impl Document {
 }
 
 pub struct ParsedQuery {
-    pub tokens: BTreeSet<String>,
+    pub tokens: Vec<Option<usize>>,
 }
 
 impl ParsedQuery {
     pub fn check_match(&self, document: &Document) -> bool {
+        if self.tokens.contains(&None) {
+            return false;
+        }
         // Check that all tokens are in document
         self.tokens
             .iter()
-            .all(|query_token| document.tokens.contains(query_token))
+            // unwrap crash safety: all tokens exist in the vocabulary if it passes the above check
+            .all(|query_token| document.tokens.binary_search(&query_token.unwrap()).is_ok())
```

Comment:
We never actually sorted `document.tokens` anywhere, binary search gives random results here. That is what causes error in tests.

Reply:
ooh :open_mouth: .. I was pulling my hair out and never checked that one. Thanks for catching it.